### PR TITLE
feat(zk): implement gen vkey

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR implements selector compression and almost completes verification key generation hopefully. Should be merged after #136 
I'll add unittests later!

### CODE REVIEW

File: tachyon/zk/plonk/circuit/selector_compressor.h [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+  using AllocateFixedColumnCallback =
+      base::RepeatingCallback<std::unique_ptr<Expression<F>>()>;
+
+  // This function takes a vector that defines each selector as well as a
+  // closure used to allocate new fixed columns, and returns the assignment of
+  // each combination as well as details about each selector assignment.
+  //
+  // This function takes
+  // * |selectors|, a vector of vector of bool.
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/selector_compressor.h:47]:
```suggestion
  // * |selectors_in|, a vector of vector of bool.
```

File: tachyon/zk/plonk/circuit/selector_compressor.h [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+          const auto& [activations, max_degree] = e;
+          return SelectorDescription(selector_index, activations, max_degree);
+        });
+
+    // The length of all provided selectors must be the same.
+    size_t n = selectors[0].activations().size();
+    DCHECK(std::all_of(selectors.begin(), selectors.end(),
+                       [n](const std::vector<bool>& activations) {
+                         return activations.size() == n;
+                       }));
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/selector_compressor.h:77]: Can't we do this check above before zipping?

**Review by dongchangYoo — APPROVED** [review]
Else, LGTM

File: tachyon/zk/plonk/circuit/expressions/expression_factory.h [hunk 3]
```diff
@@ -15,6 +15,7 @@
 #include "tachyon/zk/plonk/circuit/expressions/advice_expression.h"
 #include "tachyon/zk/plonk/circuit/expressions/challenge_expression.h"
 #include "tachyon/zk/plonk/circuit/expressions/constant_expression.h"
+#include "tachyon/zk/plonk/circuit/expressions/evaluator/selector_replacer.h"
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/circuit/expressions/expression_factory.h:18]: This is included for convenience. If you see in `expression.h`, there `ReplaceSelectors` is declared and this method is defined in separated header, `selector_replacer.h`. So without including `selector_replacer.h`, it isn't able to be built.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**